src/main/java/com/hivemq/client/mqtt/mqtt3/message/publish/Mqtt3Publish.java

```diff
@@ -77,6 +77,12 @@ public interface Mqtt3Publish extends Mqtt3Message {
      */
     boolean isRetain();
 
+    /**
+     * @return the optional local correlation data of this Publish message. This data is never propagated and kept
+     * locally for correlation.
+     */
+    Object localCorrelationData();
```

Comment:
This API needs to be discussed.

Reply:
Do you mean the usage of object (yep, could replace it with a generic but that would change general signature of the class) or something else?

Reply:
@YannickWeber any objections to merging this and then punting the improvement (that may or may not be needed) for a future date?  `Object` isn't "slick" but it serves the purpose for now without complexity.  Plus this isn't a customer requested feature.

Reply:
Yes I have objections. This is API, API can only change in major versions, therefore we need to be very cautious what we are adding and need to double check for expandability and clarity to make it future-proof.

Reply:
Well said but incorrect actually.  For clients, API can be added to in minor versions (via method overloading with defaults for example) but API removals only happen in major versions.  Your approach is correct for broker considering the many millions of devices using the API that will won't ever update.

In the client world, things aren't as strict since package updates are manual.

With this in mind, this library hasn't had a release in 2 years and this PR has been open for 4 months.  Considering the future of this space, we need agility.  And although these strict policies have good intentions, they really just end up choking progress.

So let's find the middle ground.  I approve this PR if the old call methods still works and this change is additive.  Thoughts?

Reply:
> need to double check for expandability and clarity to make it future-proof

One addition, I agree on this point completely.

Reply:
I agree with moving cautiously, just to provide some context:
The newly added API is strictly for local correlation when inside a reactive pipeline.
I used Object because it was the least invasive and most flexible in that case.
With some fighting we might turn this into a generic.
We could also generate some id for these local interactions which I think would be a little much since this must never be transferred over the wire.

Reply:
Of course we can add API in minor versions, but we can only remove methods in major versions. Therefore, I want us to be very careful about API design, as we can't easily correct potential API issues. 

> In the client world, things aren't as strict since package updates are manual.

I would argue that API stability and caution is of high importance independent of where it is used.


My suggestion would be to not return an Object but rather add an Interface `LocalCorrelationData`. So that we can expand the return value easier in the future.

Reply:
An unsafe cast can be avoided with an instanceof, which got a lot nicer with recent Java-versions.

From a type-perspective it would be best to introduce a generic, like Ractor-Kafka is doing it (https://github.com/reactor/reactor-kafka/blob/main/src/main/java/reactor/kafka/sender/SenderRecord.java#L30). 

Everything without a generic will require an unsafe cast at some point.

Reply:
A final decision would be ideal.  Let me know if there are any other thoughts.